u8timeseries/models/fft.py

```diff
@@ -0,0 +1,125 @@
+from .autoregressive_model import AutoRegressiveModel
+from ..timeseries import TimeSeries
+from ..custom_logging import raise_if_not, time_log, get_logger
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+
+logger = get_logger(__name__)
+
+
+### helper functions that might have to be relocated ###
+
+def compare_seasonality(ts_1: pd.Timestamp, ts_2: pd.Timestamp, required_matches: set):
+    """
+    Compares two timestamps according two a given set of attributes (such as minute, hour, day, etc.).
+    It returns true if and only if the two timestamps are matching in all given attributes.
+
+    :param ts_1: First timestamp that will be compared.
+    :param ts_2: Second timestamp that will be compared.
+    :required_matches: A set of pd.Timestamp attributes which ts_1 and ts_2 will be checked on.
+    :return: True if and only if 'ts_1' and 'ts_2' match in all attributes given in 'required_matches'.
+    """
+    is_matching = True
+    for required_match in required_matches:
+        is_matching = is_matching and (getattr(ts_1, required_match) == getattr(ts_2, required_match))
+    return is_matching
+
+def crop_to_match_seasons(series: 'TimeSeries', required_matches: set):
+    """
+    Crops a given TimeSeries 'series' that will be used as a training set in such
+    a way that its first entry has a timestamp that matches the first timestamp
+    right after the end of 'series' in all attributes given in 'required_matches'.
+    If no such timestamp can be found, the original TimeSeries instance is returned.
+
+    :param series: TimeSeries instance to be cropped.
+    :param ts_2: Second timestamp that will be compared.
+    :required_matches: A set of pd.Timestamp attributes which will be used to choose the cropping point.
+    :return: New TimeSeries instance that is cropped as described above.
+    """
+    first_ts = series._series.index[0]
+    freq = first_ts.freq
+    pred_ts = series._series.index[-1] + freq
+
+    # start at first timestamp of given series and move forward until a matching timestamp is found
+    curr_ts = first_ts
+    while (curr_ts < pred_ts - 4 * freq):
+        curr_ts += freq
+        if compare_seasonality(pred_ts, curr_ts, required_matches):
+            new_series = series.drop_before(curr_ts)
+            return new_series
+    
+    logger.warning("No matching timestamp could be found, returning original TimeSeries.")
+    return series
+
+class FFT(AutoRegressiveModel):
+
+    def __init__(self, filter_first_n: int = -1, required_matches: set = {}, detrend: bool = False, detrend_poly_degree: int = 3):
```

Comment:
How about renaming `filter_first_n` into `nr_freqs_to_keep` ?
Also concerning the `required_matches`, I think this will likely be quite complex to understand & use (especially because it requires knowing the pd.Timestamps attributes). How about an "auto" version, which, when predict() is called, first tries to figure out the correct phase shift?

Reply:
Also: I would rename `detrend` to `trend` and `detrend_poly_degree` to `trend_poly_degree`.
Finally, instead of default values of -1, it's probably better to use a type hint Optional[int] to make it clear this is not required (and set it to None by default).

Reply:
done